proposals/0000-local-do.rst

```diff
@@ -0,0 +1,281 @@
+
+For instance we could make a builder for monoids:
+
+::
+
+  module Data.Monoid.Builder where
+    data MonoidBuilder = MonoidBuilder
+      { (>>) :: Monoid a => a -> a -> a
+      }
+
+    builder :: MonoidBuilder
+    builder = MonoidBuilder (<>)
+
+  module X where
+    import qualified Data.Monoid.Builder as Monoid
+
+    f = do @Monoid.builder
+      Sum 2
+      Sum 3
+      Sum 5
+      Sum 8
+
+If one would try to use ``x <- u`` with ``Monoid.builder``, GHC would
+raise an error *even if there is a value for ``(>>=)`` in scope*.
+
+Importing ``-XLocalDo`` doesn't change the meaning of existing do-expressions: they will pick up the ``builder`` from the ``Prelude``, which has the same meaning as current default.
+
+``LocalDo`` interferes with ``RebindableSyntax``. We propose that ``LocalDo`` take precedence when both are enabled.
+
+The syntax was chosen to resemble that of visible type applications (as it also makes visible arguments which were previously hidden). There is no syntax conflicts, as ``do`` is not actually a function, therefore the notation ``@<expr>`` cannot occur at this site currently. This is still true after `Type applications in patterns <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0031-type-applications-in-patterns.rst>`_, even if one chooses to use the whitespace syntax *and* writes the first statement on the same line as the ``do``: no pattern can start with an ``@``.
+
+Costs and Drawbacks
+-------------------
+
+The do-expression store, during type-checking, which expression they will use for ``(>>=)``, etc… So the core infrastructure is actually already there. We anticipate the cost of implementation and maintenance of this feature to be very low.
+
+Alternatives
+------------
+
+Syntax variations
+~~~~~~~~~~~~~~~~~
+
+In the pull request discussion some concerns were raised about the use of the syntax ``do @builder``, as ``builder`` is not a type (``@`` is typically followed by a type) and because a competing idea would be to have ``do @MonadType`` be the regular ``do`` on monads, but with the monad specified.
+
+Both points can be addressed
+
+- ``@`` is better understood as something which makes an *invisible* argument visible. As it happens, invisible arguments in current Haskell are always types, but it may be best to consider it a coincidence. Especially as Haskell is moving towards more type dependency and more control over the visibility of arguments.
+- It is possible, within the changes of this proposal, to specify the monad type:
+
+  ::
+
+    do @(builder @MonadType)
+      <do block>
+
+  where ``builder`` is the standard builder from ``Control.Monad`` (in the discussion of the pull request, this ``builder`` was also referred to as ``withMonad``).
+
+Nevertheless an alternative syntax has been proposed:
+
+- ``do via <aexpr>``. It is modelled after the ``deriving via`` construction. The implications on the parser are less clear then ``@<aexpr>``, however.
```

Comment:
Would be nice if `via` was a proper keyword anyways, but it isn’t, so I find `do @foo` good enough